designs/RFC - Network Setup Teardown Optimization.md

```diff
@@ -0,0 +1,396 @@
+    comp.wait %w1 : !comp.event<ocl>
+    %2 = comp.launch_kernel %k1, %c4, %c1, %c1, %c1, %c2, %c32, %1 : (!comp.kernel<ocl, 11, 8x32xf32>, index, index, index, index, index, index, memref<8x32xf32, 11>) -> !comp.event<ocl>
+    %3 = comp.schedule_read %arg3 from %1 on %0 wait for %2 : (memref<8x32xf32>, memref<8x32xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
+    comp.wait %3 : !comp.event<ocl>
+    %w4 = comp.schedule_write %arg2 to %4 on %0 : (!comp.execenv<ocl:0,(11)>, memref<16x32xf32>, memref<16x32xf32, 11>) -> !comp.event<ocl>
+    comp.wait %w4 : !comp.event<ocl>
+    %w5 = comp.schedule_write %arg1 to %5 on %0 : (!comp.execenv<ocl:0,(11)>, memref<8x16xf32>, memref<8x16xf32, 11>) -> !comp.event<ocl>
+    comp.wait %w5 : !comp.event<ocl>
+    %6 = comp.schedule_write %arg3 to %1 on %0 : (memref<8x32xf32>, memref<8x32xf32, 11>, !comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
+    comp.wait %6 : !comp.event<ocl>
+    %7 = comp.launch_kernel %k2, %c1, %c1, %c1, %c8, %c4, %c1, %4, %5, %1 : (!comp.kernel<ocl, 11, 16x32xf32, 8x16xf32, 8x32xf32>, index, index, index, index, index, index, memref<16x32xf32, 11>, memref<8x16xf32, 11>, memref<8x32xf32, 11>) -> !comp.event<ocl>
+    %8 = comp.schedule_read %arg2 from %4 on %0 wait for %7 : (memref<16x32xf32>, memref<16x32xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
+    %9 = comp.schedule_read %arg1 from %5 on %0 wait for %7 : (memref<8x16xf32>, memref<8x16xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
+    %10 = comp.schedule_read %arg3 from %1 on %0 wait for %7 : (memref<8x32xf32>, memref<8x32xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
+    comp.wait %8, %9, %10 : !comp.event<ocl>, !comp.event<ocl>, !comp.event<ocl>
+  }
+  comp.destroy_kernel %k1 : !comp.kernel<ocl, 11, 8x32xf32>
+  comp.destroy_kernel %k2 : !comp.kernel<ocl, 11, 16x32xf32, 8x16xf32, 8x32xf32>
+  comp.dealloc %0 %4 : (!comp.execenv<ocl:0,(11)>, memref<16x32xf32, 11>) -> ()
+  comp.dealloc %0 %5 : (!comp.execenv<ocl:0,(11)>, memref<8x16xf32, 11>) -> ()
+  comp.dealloc %0 %1 : (!comp.execenv<ocl:0,(11)>, memref<8x32xf32, 11>) -> ()
+  comp.destroy_execenv %0 : !comp.execenv<ocl:0,(11)>
+  return
+}
+```
+
+### Wait Cleanup
+
+Although it's not strictly necessary for setup/teardown optimization, the code's a bit more optimal (and easier to grok)
+if we coalesce `comp.wait` operations into subsequent operations that depend on the values touched by the `comp.wait`;
+we can also optimize a bit by eliminating redundant memory transfers.
+
+(Note that since we don't have information about which values
+are kernel inputs and which are kernel outputs, kernels reading the same data will be forced to serialize; fixing this
+is an important optimization, but not one we're considering in this document).
```

Comment:
We can determine which are inputs vs outputs by doing analysis. I don't think we need a special marker from the user.

Reply:
Oh, good point -- just to clarify, do you mean by analyzing the SPIR-V code?

Reply:
Best to probably do it during PXA, because we have interfaces for determining read/write and being able to follow the chain.

Reply:
Done